### ISSUE TITLE
Configuration Binding Generator Fails to Generate Required Method

### ISSUE THREAD

**Issue by chrisoverzero** [dotnet/runtime#92771]
### Description

When configuration binding generation is enabled by means of:

```xml
<EnableConfigurationBindingGenerator>true</EnableConfigurationBindingGenerator>
<Features>$(Features);InterceptorsPreview</Features>
```

…a project can fail to compile with this error message:

```
/app/repro/Microsoft.Extensions.Configuration.Binder.SourceGeneration/Microsoft.Extensions.Configuration.Binder.SourceGeneration.ConfigurationBindingGenerator/BindingExtensions.g.cs(73,20): e
rror CS0103: The name 'AsConfigWithChildren' does not exist in the current context [/app/repro/repro.csproj]
```

### Reproduction Steps

Given the following project file:

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net8.0</TargetFramework>
    <ImplicitUsings>enable</ImplicitUsings>
    <Nullable>enable</Nullable>
    <EnableConfigurationBindingGenerator>true</EnableConfigurationBindingGenerator>
    <Features>$(Features);InterceptorsPreview</Features>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.Extensions.Hosting" Version="8.0.0-rtm.23477.9" />
    <PackageReference Include="Microsoft.Extensions.Configuration.Binder" Version="8.0.0-rtm.23477.9" />
  </ItemGroup>

</Project>
```

…and the following Program.cs:

```csharp
using Microsoft.Extensions.Configuration;

var c = new ConfigurationBuilder().Build();
var x = c.GetSection("Anything").Get<TimeSpan?>();
Console.WriteLine(x.GetValueOrDefault());
```

…run `dotnet build` via SDK version "8.0.100-rtm.23477.37".

### Expected behavior

The compilation succeeds, and the value is bound from configuration.

### Actual behavior

Compilation fails with the following error:

```
/app/repro/Microsoft.Extensions.Configuration.Binder.SourceGeneration/Microsoft.Extensions.Configuration.Binder.SourceGeneration.ConfigurationBindingGenerator/BindingExtensions.g.cs(73,20): e
rror CS0103: The name 'AsConfigWithChildren' does not exist in the current context [/app/repro/repro.csproj]
```

This will occur for… I'm not sure how to characterize it. It will mostly occur for any type which is value-like in nature – not necessarily a _value type_, that is. `string` and `TimeSpan` (as well as `TimeSpan?`) fail, but `System.Math.Complex` and user-defined types with properties succeed. Except user-defined structs go back to failing when nullable.

### Regression?

This succeeded back in the RC1 period, as [here](https://github.com/dotnet/runtime/issues/86557). But that was before interceptors.

### Known Workarounds

Since the presence of a property-bearing type causes `AsConfigWithChildren` to be generated, one can bind to a value of such a type and discard it in a no-oppish way.

```csharp
using Microsoft.Extensions.Configuration;

var c = new ConfigurationBuilder().Build();
_ = c.GetSection("Anything").Get<System.Math.Complex>();
var x = c.GetSection("Anything").Get<TimeSpan?>();
Console.WriteLine(x.GetValueOrDefault());
```

### Configuration

.NET Version: 8.0.100-rtm.23477.37
OS and version: macOS 13.5.1 (22G90)
Architecture: x64
Specific?: No, I don't think so.
Blazor: ❌

**Comment by tarekgh** [comment]
CC @layomia @ericstj

**Comment by ericstj** [comment]
I can reproduce on the 8.0 bits, does not reproduce on 9.0 as of this change: https://github.com/dotnet/runtime/commit/8e760396b1be0e9094634a3e77741bf2408f1d74

We are planning to port that into 8.0 so it should provide the fix.

I was also able to workaround this by adding other non-primitive bind calls in the same assembly.

**Comment by ericstj** [comment]
This is fixed in main, and now fixed in 8.0 with https://github.com/dotnet/runtime/pull/92730
